### PR TITLE
fix(lint/useTemplate): correctly handles parenthesized expressions and respects type coercions

### LINKED ISSUE #1192: 💅 `useStringLiterals` produce false recommendations
### Environment information

```bash
CLI:
  Version:                      1.4.0
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "alacritty"
  JS_RUNTIME_VERSION:           "v20.10.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.2.3"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### Rule name

useStringLiterals

### Playground link

N/A

### Expected result

For the following code:

```
function captureRandomMessage() {
  Sentry.captureMessage('Message no ' + (Date.now() + Math.random()));
}
```

It produces:

Screenshot 2023-12-14 at 10 53 24 AM

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

### LINKED ISSUE #1456: 💅 `useTemplate` doesn't report expressions with an interpolated template literal and non-string expressions
### Environment information

```bash
N/A
```

### Rule name

useTemplate

### Playground link

https://biomejs.dev/playground/?code=LwAvACAARABvAGUAcwBuACcAdAAgAHIAZQBwAG8AcgB0ADoACgBgAGEAJAB7ADEAfQBgACAAKwAgADIAOwAKAAoALwAvACAAVwBoAGkAbABlACAAcgBlAHAAbwByAHQAaQBuAGcAOgAKACIAYQAiACAAKwAgADEAOwAKAGAAYQBgACAAKwAgADEAOwAKAGAAYQAkAHsAMQB9AGAAIAArACAAIgBiACIAOwAKAA%3D%3D

### Expected result

The following code should be reported:

```js
`a${1}` + 2;
```

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

### PR DESCRIPTION
**Pull request by n-gude**
## Summary

Closes #1192.

[useTemplate](https://biomejs.dev/linter/rules/use-template/) now correctly handles parenthesized expressions and respects type coercions.

It also fixes cases where the string literal was 'too deeply' nested, e.g.:
```js
1 + (2 + "a")
```

```diff,js
-`${2}a`
+`${1}${2}a`
```

Closes #1456.

[useTemplate](https://biomejs.dev/linter/rules/use-template/) now reports expressions with an interpolated template literal and non-string expressions.

## Test Plan

Tests included.

### CODE REVIEW

File: crates/biome_js_analyze/src/analyzers/style/use_template.rs [hunk 1]
```diff
@@ -153,36 +153,43 @@ fn template_expression_from_binary_expression(
     // While `template_elements` is empty, we keep track of the last left node.
     // Once we see the first string/template literal,
     // we insert `last_left_node` in `template_elements` and the seen string/template literal.
-    // ANy subsequent expression is directly inserted in `template_elements`.
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/style/use_template.rs:156]: These comments should be updated.

File: crates/biome_js_analyze/src/analyzers/style/use_template.rs [hunk 2]
```diff
@@ -153,36 +153,43 @@ fn template_expression_from_binary_expression(
...
                 match node {
-                    AnyJsExpression::JsParenthesizedExpression(_) => continue,
-                    AnyJsExpression::JsBinaryExpression(node)
-                        if matches!(node.operator().ok()?, JsBinaryOperator::Plus) =>
+                    AnyJsExpression::JsParenthesizedExpression(_) => {
+                        continue;
+                    }
+                    AnyJsExpression::JsBinaryExpression(ref binary)
+                        if matches!(binary.operator().ok()?, JsBinaryOperator::Plus)
+                            && evaluates_to_string(binary)? =>
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/style/use_template.rs:170]: Similarly to `template_expression_from_binary_expression`, `evaluates_to_string` traverses the subtree. This duplicates some traversals. Do you think we could avoid double traversal?

**Review by Conaclos — COMMENTED** [review]
Looks good to me! I left suggestions/questions :)

**Comment by ematipico** [comment]
Hi @n-gude, are you still interested in completing the PR?

**Comment by n-gude** [comment]
@ematipico Yeah, I'm still trying to get one edge case right after rewriting the logic to avoid the double traversals. There were in fact a lot of double traversals...

**Comment by ematipico** [comment]
Let us know if you need any help

**Comment by n-gude** [comment]
Thanks, but I finally got it!

I noticed that the rule doesn't report expressions with an interpolated template literal and non-string expressions:
```js
`a${1}` + 2;
```
Because the equivalent ESLint rule reports that case, I assume this is a bug, right?
I created a fix for that. Is it ok to put it in this PR or should I create another for that?

**Comment by Conaclos** [comment]
> Thanks, but I finally got it!
> 
> I noticed that the rule doesn't report expressions with an interpolated template literal and non-string expressions:
> 
> ```js
> `a${1}` + 2;
> ```
> 
> Because the equivalent ESLint rule reports that case, I assume this is a bug, right? I created a fix for that. Is it ok to put it in this PR or should I create another for that?

It is ok to include the fix. Also, It could be nice to open an issue to keep some trace of this.

File: crates/biome_js_analyze/src/analyzers/style/use_template.rs [hunk 3]
```diff
@@ -150,51 +157,94 @@ fn can_be_template_literal(node: &JsBinaryExpression) -> Option<bool> {
...
+                    AnyJsExpression::JsBinaryExpression(ref binary)
+                        if binary.operator().ok()? == JsBinaryOperator::Plus =>
                     {
+                        left_expressions_stack.push(vec![]);
+                        binary_evaluates_to_string_stack.push(false);
                         continue;
                     }
+                    _ => {}
+                }
+                iter.skip_subtree();
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/style/use_template.rs:185]: By moving `iter.skip_subtree()` in the match, we could remove the previous `continue` statements?
```suggestion
                    _ => iter.skip_subtree(),
                }
```

File: crates/biome_js_analyze/src/analyzers/style/use_template.rs [hunk 4]
```diff
@@ -150,51 +157,94 @@ fn can_be_template_literal(node: &JsBinaryExpression) -> Option<bool> {
...
+            }
+            WalkEvent::Leave(node) => {
+                let expression = AnyJsExpression::cast(node)?;
+                match expression {
+                    // Skip parenthesized expressions, because they would be added twice to
+                    // `left_expressions_stack` or `template_elements` (see the last match arm):
+                    // First the contained expression and then the parenthesized expression itself.
+                    AnyJsExpression::JsParenthesizedExpression(_) => continue,
+                    AnyJsExpression::JsBinaryExpression(ref binary)
+                        if binary.operator().ok()? == JsBinaryOperator::Plus =>
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/style/use_template.rs:195]:
```suggestion
                        if binary.operator() == Ok(JsBinaryOperator::Plus) =>
```

File: crates/biome_js_analyze/src/analyzers/style/use_template.rs [hunk 5]
```diff
@@ -150,51 +157,94 @@ fn can_be_template_literal(node: &JsBinaryExpression) -> Option<bool> {
...
+                        binary_evaluates_to_string_stack.push(false);
                         continue;
                     }
+                    _ => {}
+                }
+                iter.skip_subtree();
+            }
+            WalkEvent::Leave(node) => {
+                let expression = AnyJsExpression::cast(node)?;
+                match expression {
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/analyzers/style/use_template.rs:189]: Using `&` allows removing the `ref` in following match patterns.

```suggestion
                match &expression {
```

**Review by Conaclos — APPROVED** [review]

**Comment by Conaclos** [comment]
Thanks for this fantastic contribution! This is just in time for the upcoming release ❤️

**Comment by n-gude** [comment]
@Conaclos Thank you for these detailed reviews! They really help me catch those things myself in the future (hopefully 😉).

**Pull request merged**